### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by efc-michaelwiezik**
Hey,

it's initial version of the project, hence all development done directly in develop branch.
It contains only password validation extracted from the one-big-app module.

Please let me know in case of any questions.

Thanks & Regards,

Michael

### CODE REVIEW

File: src/main/java/com/efinancialcareers/utilities/validation/PasswordValidatorImpl.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+        this.passwordStrengthCalculator = passwordStrengthCalculator;
+        this.passwordConstraintsValidator = passwordConstraintsValidator;
+
+    }
+
+    @Override
+    public void validatePassword(@NotNull String password) {
+
+        if (password == null) {
```
Comment (efc-carloscornejo) [inline src/main/java/com/efinancialcareers/utilities/validation/PasswordValidatorImpl.java]: if you make this parameter @NotNull, password==null will never happen. This is the contract you are creating with the annotation. So either make password @Nullable or remove the password==null if statement.

File: src/main/java/com/efinancialcareers/utilities/validation/password/PasswordConstraintsValidator.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+    public static String SPACE_NOT_ALLOWED_MESSAGE = "Space is not allowed in password";
+
+    public static String INCORRECT_PASSWORD_LENGTH = "Password has to be at least " + MIN_PASSWORD_LENGTH + " and " +
+            "at most " + MAX_PASSWORD_LENGTH + " characters";
+
+    public static String NO_SPECIAL_CHARACTERS_MESSAGE = "Password has to contain at least one character " +
+            "from the list: " + PUNCTUATION_CHARS;
+
+    public void validatePassword(String password) {
```
Comment (efc-carloscornejo) [inline src/main/java/com/efinancialcareers/utilities/validation/password/PasswordConstraintsValidator.java]: Yo are not using jsr305 annotations as in the other cases, is this expected??

File: src/main/java/com/efinancialcareers/utilities/validation/password/PasswordConstraintsValidator.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+            MIN_PASSWORD_LENGTH = 8,
+            MAX_PASSWORD_LENGTH = 40;
+
+    private static final Pattern
+            INVALID_CHARS = Pattern.compile("@|&|<|-|=|'"),
+            CONTAINS_SPACE = Pattern.compile("\\s"),
+            ALL_CHARS = Pattern.compile("\\p{Lower}|\\p{Upper}"),
+            DIGIT_CHARS = Pattern.compile("\\p{Digit}"),
+            PUNCTUATION_CHARS = Pattern.compile("!|#|\\$|%|\\^|\\*|\\?|_|~");
+
```
Comment (efc-carloscornejo) [inline src/main/java/com/efinancialcareers/utilities/validation/password/PasswordConstraintsValidator.java]: Are all these guys meant to be public?? is someone externally using them? If that is the case I'd move it to the non-existing interface. BTW why there is not interface :-) as you are you are exposing public methods

File: src/main/java/com/efinancialcareers/utilities/validation/password/PasswordStrengthCalculator.java [hunk 4]
```diff
@@ -0,0 +1,120 @@
+package com.efinancialcareers.utilities.validation.password;
+
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
```
Comment (efc-carloscornejo) [inline src/main/java/com/efinancialcareers/utilities/validation/password/PasswordStrengthCalculator.java]: No interface... it should have

File: src/main/java/com/efinancialcareers/utilities/validation/password/PasswordStrengthCalculator.java [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+public class PasswordStrengthCalculator {
+
+    private static final Pattern
+            LOWER_CASE_CHARS = Pattern.compile("\\p{Lower}"),
+            UPPER_CASE_CHARS = Pattern.compile("\\p{Upper}"),
+            DIGIT_CHARS = Pattern.compile("\\p{Digit}"),
+            SPECIAL_CHARS = Pattern.compile("!|#|\\$|%|\\^|\\*|\\?|_|~"),
+            ALL_CHARS = Pattern.compile("\\p{Lower}|\\p{Upper}");
+
+
```
Comment (efc-carloscornejo) [inline src/main/java/com/efinancialcareers/utilities/validation/password/PasswordStrengthCalculator.java]: are we missing again jsr305 annotations?

File: src/main/java/com/efinancialcareers/utilities/validation/PasswordValidatorImpl.java [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+    }
+
+    public PasswordValidatorImpl(PasswordStrengthCalculator passwordStrengthCalculator,
+                                 PasswordConstraintsValidator passwordConstraintsValidator) {
+
+        this.passwordStrengthCalculator = passwordStrengthCalculator;
+        this.passwordConstraintsValidator = passwordConstraintsValidator;
+
+    }
+
```
Comment (efc-carloscornejo) [inline src/main/java/com/efinancialcareers/utilities/validation/PasswordValidatorImpl.java]: If you make @NotNull String password then you don't need password == null. Either make it @Nullable or delete the if

**Comment by efc-carloscornejo** [comment]
Please have a look to the comment I added and let me know your thoughts.

Cheers,
Carlos M.

File: src/main/java/com/efinancialcareers/utilities/validation/password/PasswordConstraintsValidator.java (hunk 2 above)
Comment (efc-michaelwiezik) [inline src/main/java/com/efinancialcareers/utilities/validation/password/PasswordConstraintsValidator.java]: Added NotNull with Nonnull annotations. However, none of them replaces runtime assertions as Nonnull does only static check and NotNull doesn't result with exception without additional configuration.

File: src/main/java/com/efinancialcareers/utilities/validation/password/PasswordConstraintsValidator.java (hunk 3 above)
Comment (efc-michaelwiezik) [inline src/main/java/com/efinancialcareers/utilities/validation/password/PasswordConstraintsValidator.java]: MIN_PASSWORD_LENGTH and MAX_PASSWORD_LENGTH made private - my bad.

File: src/main/java/com/efinancialcareers/utilities/validation/password/PasswordStrengthCalculator.java (hunk 4 above)
Comment (efc-michaelwiezik) [inline src/main/java/com/efinancialcareers/utilities/validation/password/PasswordStrengthCalculator.java]: Found this article as a rationale against 1-2-1 interface-implementation pairing, if you have time to read, please let me know your feedback. http://martinfowler.com/bliki/InterfaceImplementationPair.html

File: src/main/java/com/efinancialcareers/utilities/validation/password/PasswordStrengthCalculator.java (hunk 5 above)
Comment (efc-michaelwiezik) [inline src/main/java/com/efinancialcareers/utilities/validation/password/PasswordStrengthCalculator.java]: Added @Nonnull to public methods.

File: src/main/java/com/efinancialcareers/utilities/validation/PasswordValidatorImpl.java (hunk 6 above)
Comment (efc-michaelwiezik) [inline src/main/java/com/efinancialcareers/utilities/validation/PasswordValidatorImpl.java]: As above.

**Pull request merged**